src/controller/CHIPDeviceController.cpp

```diff
@@ -264,54 +278,61 @@ void DeviceController::OnLocalNOCChainGeneration(void * context, CHIP_ERROR stat
     }
 }
 
-CHIP_ERROR DeviceController::LoadLocalCredentials(Transport::FabricInfo * fabric)
+CHIP_ERROR DeviceController::GenerateLocalCredentials()
 {
-    ChipLogProgress(Controller, "Getting operational keys");
-    Crypto::P256Keypair * keypair = fabric->GetOperationalKey();
+    ChipLogProgress(Controller, "Generating new local credentials");
```

Comment:
For most real applications, this will actually need to be a custom step that may interface with a CA and can't directly be done locally.

Let's discuss the cleanest way to do this without too many changes. I think that this could be resolved by:

1. Making `GenerateLocalCredentials` an "example building block"
2. Providing the necessary API surface to `CHIPDeviceController` so that the outcome `GenerateLocalCredentials` (e.g. an NOC chain and operational key pair) can be added to the storage used by the stack to refer to credentials wherever they are needed.

Bear in mind that, asymptotically, to be spec-compliant, **all Nodes will need to be "controllers" of other Nodes**. The more assumptions we have about CHIPDeviceController code being "special" non-device code that manages credentials differently than a "device", the more pain it will be to do the following features:

- OTA requestor support (required by most devices)
- Time synchronization cluster support
- True application of bindings (e.g. sensor actuating an actuator).

I think we are at the point where a bit of additional refactor at this juncture will save a lot of time later.

Reply:
Yes, this is important.  We need the sdk to allow us to initialize the stack with operational credentials acquired by the app by any method it chooses.  It is fundamental that there is always a 'first node' on each fabric that was not commissioned by other nodes and will have bootstrapped itself onto the fabric by means not defined in the core spec.  The controller currently has a notion of this, but presumes it should be the entity that bootstraps the stack onto a fabric for this case.

That's one way to do things, but can't be the only way.  The code needs to support a notion of op creds that exist before the stack ever starts.  As things are currently stand, this implies providing an interface to simply pass in an `OperationalCredentialSet` object that was instantiated by means that do not need to be defined here.  The contents of this object need to be defined, but how it comes to exist does not.  At that point, something like `GenerateLocalCredentials` would become part of an example application, not the core sdk.

Reply:
Please open an issue on this for a subsequent PR.

Reply:
This issue should cover it: https://github.com/project-chip/connectedhomeip/issues/8910